openfeature/client_test.go

```diff
@@ -764,6 +764,148 @@ func TestRequirement_1_4_13(t *testing.T) {
+		t.Error("client returned by NewClient doesn't implement the 1.6.* requirements interface")
+	}
+}
+
+// Requirement_6_1_3
+// The evaluation context passed to the provider's track function MUST be merged in the order, with duplicate values being overwritten:
+// - API (global; lowest precedence)
+// - transaction
+// - client
+// - invocation (highest precedence)
+
+// Requirement_6_1_4
+// If the client's `track` function is called and the associated provider does not implement tracking, the client's `track` function MUST no-op.,
+func TestForTracking(t *testing.T) {
+	type inputCtx struct {
+		api        EvaluationContext
+		txn        EvaluationContext
+		client     EvaluationContext
+		invocation EvaluationContext
+	}
+
+	ctrl := gomock.NewController(t)
+	mockNonTrackingProvider := NewMockFeatureProvider(ctrl)
+	mockNonTrackingProvider.EXPECT().Metadata().AnyTimes()
+
+	// mockTrackingProvider is a feature provider that implements tracking handler contract.
+	mockTrackingProvider := struct {
+		*MockTrackingHandler
+		*MockFeatureProvider
+	}{
+		MockTrackingHandler: NewMockTrackingHandler(ctrl),
+		MockFeatureProvider: NewMockFeatureProvider(ctrl),
+	}
+
+	mockTrackingProvider.MockFeatureProvider.EXPECT().Metadata().AnyTimes()
+
+	tests := map[string]struct {
+		inCtx         inputCtx
+		inputProvider FeatureProvider
+		outCtx        EvaluationContext
+		outputHandler TrackingHandler
+	}{
+		"merging in correct order": {
+			inCtx: inputCtx{
+				api: EvaluationContext{
+					attributes: map[string]interface{}{
+						"1": "api",
+						"2": "api",
+						"3": "api",
+						"4": "api",
+					},
+				},
+				txn: EvaluationContext{
+					attributes: map[string]interface{}{
+						"2": "txn",
+						"3": "txn",
+						"4": "txn",
+					},
+				},
+				client: EvaluationContext{
+					attributes: map[string]interface{}{
+						"3": "client",
+						"4": "client",
+					},
+				},
+				invocation: EvaluationContext{
+					attributes: map[string]interface{}{
+						"4": "invocation",
+					},
+				},
+			},
+			inputProvider: mockNonTrackingProvider,
+			outCtx: EvaluationContext{
+				attributes: map[string]interface{}{
+					"1": "api",
+					"2": "txn",
+					"3": "client",
+					"4": "invocation",
+				},
+			},
+			outputHandler: NoopProvider{},
+		},
+		"non tracking provider": {
+			inputProvider: mockNonTrackingProvider,
+			outputHandler: NoopProvider{},
+		},
+		"expect no-op provider": {
+			inputProvider: mockTrackingProvider,
+			outputHandler: mockTrackingProvider,
+		},
+	}
+
+	for name, test := range tests {
+		t.Run(name, func(t *testing.T) {
+			// arrange
+			client := NewClient("test-client")
+			_ = client.api.SetProviderAndWait(test.inputProvider)
+
+			client.evaluationContext = test.inCtx.client
+			client.api.SetEvaluationContext(test.inCtx.api)
+			ctx := WithTransactionContext(context.Background(), test.inCtx.txn)
+
+			// action
+			handler, evalCtx := client.ForTracking(ctx, test.inCtx.invocation)
```

Comment:
Related to https://github.com/open-feature/go-sdk/pull/297/files#r1813416916, I think we should mostly be testing `Track` not `ForTracking`, but please let me know if I've misunderstood.

Reply:
In my opinion, It is easier to write test for the requirement `Requirement_6_1_3` and `Requirement_6_1_4` if we test `ForTracking`.

However, the requirement also states that we want to test the Track function and the evalContext passed to the `Provider Tracker`, so I have updated to test method `Track`.